adding cloud providers choice

choldgraf opened the pull request:
this adds a short section on choosing cloud providers and navigating the different services that they often offer.

choldgraf committed:
adding cloud providers choice

yuvipanda commented on content/deploy/choose_cloud.md:
```diff
@@ -0,0 +1,70 @@
+
+The first thing you'll decide is *where* your JupyterHub will run. The first
+question to ask is "will I run my JupyterHub on my own hardware, or will I
+run it in the cloud?" Running a JupyterHub on your own hardware is your best
```
IMO, we shouldn't recommend this as a best choice. Running on your own hardware is going to be always problematic when stuff breaks. Here's a list of things that might happen to your hardware that are hard to fix without more engineering, while very easy to on the cloud:

1. You need more RAM / CPU
2. External public IPs (this is surprisingly hard with your own hardware)
3. Backups (snapshots are easy in all providers)
4. You screw up your box, and need to start over.

All these will probably happen at some point, and when most people think they have hardware support they're always going to underestimate these costs.

We should recommend they run on a cloud provider unless they can't, in which case they should run on their own hardware.

yuvipanda commented on content/deploy/choose_cloud.md:
```diff
@@ -0,0 +1,70 @@
+> **An aside on using free hosted services vs. creating your own deployments**. Many cloud
+  providers offer a "free" and customized notebook service for basic usage. While these are
+  appealing on their surface, they provide a number of challenges down the road, such as
+  getting your students used to the vendor-specific interface rather than an open-source alternative.
```
```suggestion
  getting your students used to the vendor-specific interface rather than an open-source version.
```

yuvipanda commented on content/deploy/choose_cloud.md:
```diff
@@ -0,0 +1,70 @@
+
+  * **Pros**: You can get a Jupyter-like experience extremely quickly. In addition, these services are
+    often "free" (at least initially).
+  * **Cons**: This greatly increases your dependence on a single cloud provider, and the interface +
```
A big con is also that you don't have control over a lot of things there. Here's a small list:

1. Environment might not have the versions of packages you want installed (this is more a problem in Colab maybe? Unclear)
2. You don't have access to extensions or customizations as you want. No nbgitpuller, no canvas / LMS integration, no customizable resources, etc
3. ...

While the core reasons are one of control, we should definitely list some strong technical reasons too, despite the possibility they are gonna get out of date quickly. I also don't know how many of these are true right now.

yuvipanda reviewed (changes requested):
This is a good start! How should we approach recommending particular providers? I personally think you should just use Digital Ocean unless you have access to cloud credits, but I don't think that can be a 'project' opinion...

Do you think we should mention we follow this practice at Berkeley? Or should that be someplace else?

Also, thank you very much for working on this

choldgraf committed:
Update content/deploy/choose_cloud.md

Co-Authored-By: Yuvi Panda <yuvipanda@gmail.com>

lheagy commented on content/deploy/choose_cloud.md:
```diff
@@ -0,0 +1,70 @@
+
+## Choosing a cloud provider and service
+
+Cloud providers (e.g. Microsoft Azure, Google Cloud, or Digital Ocean) are where you'll
```
AWS too?

lheagy commented on content/deploy/choose_cloud.md:
```diff
@@ -0,0 +1,70 @@
+  getting your students used to the vendor-specific interface rather than an open-source version.
+  The Jupyter Project (and the Data 8 approach) aims to maximize your choice and flexibility in
+  how and where you deploy Jupyter environments on shared infrastructure, and for this reason
+  we do not recommend using most fully-managed "free" services.
```
I would suggest framing this similar to the others with "pros" and "cons". For under-resourced instructors, using a managed "free" service may be the most feasible approach for them in the first deployment of a course. We don't need to spend much time discussing how to do this, and focus in on DIY approaches, but I think there are quite a number of instructors for whom launching their own hub is a complete non-starter. If they can get started on a service, and then transition, perhaps in the next delivery of the course to using their own deployment, then that is still a net-win.

lheagy commented on content/deploy/choose_cloud.md:
```diff
@@ -0,0 +1,70 @@
+    be very similar for each.
+  * **Cons**: There is more manual setup and maintenance for you to perform. Because the resources
+    are a blank slate, you'll need to take the steps to set everything up.
+* **You use a "push-button" deployment to deploy specific resources**. Another possibility is that
```
Does this exist anywhere yet? I haven't seen a "push button" approach by any of the cloud providers yet. Are there examples we could point to? Otherwise, this discussion is quite abstract and not necessarily helpful for an instructor looking to get started.

lheagy commented on content/deploy/choose_cloud.md:
```diff
@@ -0,0 +1,70 @@
+* **You use a JupyterHub-like hosted service of a cloud provider**. The final option is that you
+  use a fully-managed service of a cloud provider, such as the Google Collaboratory, or Microsoft
+  Azure Notebooks. These often look and behave like Jupyter Notebooks, though they are often
+  built on some proprietary technology from these companies. This should be your least-preferable
```
The language here is potentially quite off-putting, "This should be your ..." is essentially imposing your value-system on instructors, and potentially comes across as judgemental if their priority system doesn't align with yours. I would recommend that we re-work the language to be a bit more matter-of-fact, and let the pros / cons do the work of providing a structure for forming an opinion.

choldgraf committed:
softening cloud language

choldgraf commented:
ok, I removed some of the more "opinionated" statements from the "choosing cloud resources" page and update the list there a bit, also added a section elsewhere on berkeley's approach to open source. Would love to hear what @lheagy and @yuvipanda think about it!

lheagy commented on content/tech/considerations.md:
```diff
@@ -6,9 +6,73 @@ doing non-trivial computational work is a big technical challenge. This page det
+contribute to it, or even modify it if they wish. The language isn't "owned" by anybody, which
+we believe is an important foundation to use for introducing students to data science.
+
+### The scipy stack - an open source ecosystem
```
```suggestion
### The Scipy stack - an open source ecosystem
```
?

lheagy reviewed (approved):
This looks really solid @choldgraf! The separation of the `Open source strategy` from the steps and decision points of an individual instructor relieves the concerns I had about the previous version. Thanks for pushing on this - it will be a valuable resource for the community!

lheagy commented:
One other quick question: do we have a pointer somewhere to guide people when to use TLJH vs. Z2JH? giving a rough "if your class is < 100?" (cc @yuvipanda) - then deploying TLJH is recommended (v. easy on digital ocean, for example), and if larger, then Z2JH. This might already be somewhere else, in the docs?

choldgraf committed:
clarifying jupyterhub distribution choice

choldgraf commented:
Good question - we do have that covered in the next section here: http://data8.org/zero-to-data-8/deploy/setup_jupyterhub.html#deploying-jupyterhub-in-the-cloud

I just pushed a commit that tries to clean up the section header and make it a bit more clear what the section is about.

@yuvipanda are you happy with the language right now? (you've got a "change requested" flag on this PR still)

choldgraf committed:
Update content/tech/considerations.md

Co-Authored-By: Lindsey Heagy <lindseyheagy@gmail.com>

choldgraf commented:
@yuvipanda I believe that I addressed your issue before, so I'm gonna merge this one and feel free to let me know if you' like me to change something moving forward!
